src/modules/ft/data_provider/history.rs

```diff
@@ -1,113 +1,199 @@
 use crate::modules::ft;
 use crate::{db_helpers, errors, types};
-use num_traits::Signed;
+use num_traits::{Signed, ToPrimitive};
 use sqlx::types::BigDecimal;
 use std::str::FromStr;
 
-// TODO PHASE 2 pagination by artificial index added to assets__fungible_token_events
-// TODO PHASE 2 change RPC call to DB call by adding absolute amount values to assets__fungible_token_events
-// TODO PHASE 2 make the decision about separate FT/MT tables or one table. Pagination implementation depends on this
 pub(crate) async fn get_ft_history(
```

Comment:
The logic of this function is really tricky.
I've left comments to help the readers, see below, and I also covered it with tests as completely as possible.
It works well, I can guarantee that. Unfortunately, without absolute balances in the DB, we can't make it simpler.